File: Runtime/TSTTransitionGroupMap.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.Common

Friend Class TstTransitionGroupMap
    Inherits STSimMapBase

    Private m_Map As New MultiLevelKeyMap4(Of TstTransitionGroup)

    Public Sub New(ByVal scenario As Scenario)
        MyBase.New(scenario)
    End Sub

    Public Function GetGroup(
        ByVal transitionTypeId As Integer,
        ByVal stratumId As Integer,
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal tertiaryStratumId As Nullable(Of Integer)) As TstTransitionGroup

        Return Me.m_Map.GetItem(
            transitionTypeId,
            stratumId,
            secondaryStratumId,
            tertiaryStratumId)

    End Function

    Public Sub AddGroup(
        ByVal transitionTypeId As Integer,
        ByVal stratumId As Nullable(Of Integer),
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal tertiaryStratumId As Nullable(Of Integer),
        ByVal item As TstTransitionGroup)

        Dim v As TstTransitionGroup = Me.m_Map.GetItemExact(transitionTypeId, stratumId, secondaryStratumId, tertiaryStratumId)

        If (v IsNot Nothing) Then

            Dim template As String =
                "A duplicate Time-Since-Transition Group was detected: More information:" & vbCrLf &
                "Transition Type={0}, {1}={2}, {3}={4}, {5}={6}."

            ExceptionUtils.ThrowArgumentException(
                template,
                Me.GetTransitionTypeName(transitionTypeId),
                Me.PrimaryStratumLabel,
                Me.GetStratumName(stratumId),
                Me.SecondaryStratumLabel,
                Me.GetSecondaryStratumName(secondaryStratumId),
                Me.TertiaryStratumLabel,
                Me.GetTertiaryStratumName(tertiaryStratumId))

        End If

        Me.m_Map.AddItem(transitionTypeId, stratumId, secondaryStratumId, tertiaryStratumId, item)
        Me.SetHasItems()

    End Sub

End Class

File: src/Runtime/TSTTransitionGroupMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.Core;
using SyncroSim.Common;

namespace SyncroSim.STSim
{
    internal class TstTransitionGroupMap : STSimMapBase
    {
        private MultiLevelKeyMap4<TstTransitionGroup> m_Map = new MultiLevelKeyMap4<TstTransitionGroup>();

        public TstTransitionGroupMap(Scenario scenario) : base(scenario)
        {
        }

        public TstTransitionGroup GetGroup(int transitionTypeId, int stratumId, int? secondaryStratumId, int? tertiaryStratumId)
        {
            return this.m_Map.GetItem(transitionTypeId, stratumId, secondaryStratumId, tertiaryStratumId);
        }

        public void AddGroup(int transitionTypeId, int? stratumId, int? secondaryStratumId, int? tertiaryStratumId, TstTransitionGroup item)
        {
            TstTransitionGroup v = this.m_Map.GetItemExact(transitionTypeId, stratumId, secondaryStratumId, tertiaryStratumId);

            if (v != null)
            {
                string template = "A duplicate Time-Since-Transition Group was detected: More information:" + Environment.NewLine + "Transition Type={0}, {1}={2}, {3}={4}, {5}={6}.";
                ExceptionUtils.ThrowArgumentException(template, this.GetTransitionTypeName(transitionTypeId), this.PrimaryStratumLabel, this.GetStratumName(stratumId), this.SecondaryStratumLabel, this.GetSecondaryStratumName(secondaryStratumId), this.TertiaryStratumLabel, this.GetTertiaryStratumName(tertiaryStratumId));
            }

            this.m_Map.AddItem(transitionTypeId, stratumId, secondaryStratumId, tertiaryStratumId, item);
            this.SetHasItems();
        }
    }
}
